### DISCUSSION TITLE
Webpack error with v6: `UnhandledSchemeError: Reading from "node:crypto" is not handled by plugins`

### DISCUSSION THREAD

**Discussion by 223740** [openpgpjs/openpgpjs#1809] [General]
Thank you for making openpgp.js available. I have encountered a bug when trying to use it that I have been unable to resolve.
`ERROR in ./node_modules/openpgp/dist/node/openpgp.mjs 4:0-39 Module not found:
Error: Can't resolve 'module' in 'C:\Users\<User>\Documents\Projekte\mail-buddy\node_modules\openpgp\dist\node' ERROR in node:crypto Module build failed: UnhandledSchemeError: Reading from "node:crypto" is not handled by plugins (Unhandled scheme). Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.`

OpenPGP.js version: 6
Affected platform (Browser or Node.js version): Node v22.9.0
I installed the module via npm, and tries to use it (I am using the webpack bundler, if this is important).

**Comment by larabr** [comment]
How have you configured webpack? For some reason it's bringing in the Node version of the lib instead of the browser one.

**Comment by 223740** [comment]
Like this. Thanks for the quick response
```
/* eslint-disable no-undef */

const devCerts = require("office-addin-dev-certs");
const CopyWebpackPlugin = require("copy-webpack-plugin");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const path = require("path");

const urlDev = "https://localhost:3000/";
const urlProd = "https://url"; // CHANGE THIS TO YOUR PRODUCTION DEPLOYMENT LOCATION

async function getHttpsOptions() {
  const httpsOptions = await devCerts.getHttpsServerOptions();
  return { ca: httpsOptions.ca, key: httpsOptions.key, cert: httpsOptions.cert };
}

// eslint-disable-next-line max-lines-per-function
module.exports = async (env, options) => {
  const dev = options.mode === "development";
  const config = {
    devtool: "source-map",
    entry: {
      polyfill: ["core-js/stable", "regenerator-runtime/runtime"],
      file: ["./path/to/file.ts", "./path/to/file.html"],
    },
    output: {
      clean: true,
    },
    resolve: {
      extensions: [".ts", ".tsx", ".html", ".js"],
    },
    module: {
      rules: [
        {
          test: /\.ts$/,
          exclude: /node_modules/,
          use: {
            loader: "babel-loader",
            options: {
              presets: ["@babel/preset-typescript"],
            },
          },
        },
        {
          test: /\.tsx?$/,
          exclude: /node_modules/,
          use: ["ts-loader"],
        },
        {
          test: /\.html$/,
          exclude: /node_modules/,
          use: "html-loader",
        },
        {
          test: /\.(png|jpg|jpeg|ttf|woff|woff2|gif|ico)$/,
          type: "asset/resource",
          generator: {
            filename: "assets/[name][ext][query]",
          },
        },
        {
          test: /\.css$/i,
          include: path.resolve(__dirname, "src"),
          use: ["style-loader", "css-loader", "postcss-loader"],
       },
      ],
    },
    plugins: [
      new HtmlWebpackPlugin({
        filename: "file.html",
        template: "./path/to/file.html",
        chunks: ["polyfill", "file"],
      }),
      new CopyWebpackPlugin({
        patterns: [
          {
            from: "assets/*",
            to: "assets/[name][ext][query]",
          },
          {
            from: "manifest*.xml",
            to: "[name]" + "[ext]",
            transform(content) {
              if (dev) {
                return content;
              } else {
                return content.toString().replace(new RegExp(urlDev, "g"), urlProd);
              }
            },
          },
        ],
      }),
    ],
    devServer: {
      headers: {
        "Access-Control-Allow-Origin": "*",
      },
      server: {
        type: "https",
        options: env.WEBPACK_BUILD || options.https !== undefined ? options.https : await getHttpsOptions(),
      },
      port: process.env.npm_package_config_dev_server_port || 3000,
    },
  };

  return config;
};
```

**Reply by larabr** [reply]
I don't have the time capacity to debug your Webpack setup, but to fix this you should investigate why it's loading the Node version of the lib (note that in v6 we have added `exports` entrypoints in the package.json, but the legacy entrypoints from v5 are also still in place; however, the lib is now declared as `module`). This specific `node:crypto` error could be fixed with workarounds, but after that you'll bump into other issues. Really it's not an option to "convert" the Node build to run in browser; we generate platform-specific builds on purpose, for optimization but also security 🙂

**Reply by 223740** [reply]
Ok, thank you nevertheless for the pointers :) Have a great day

**Reply by RaenonX** [reply]
> I don't have the time capacity to debug your Webpack setup, but to fix this you should investigate why it's loading the Node version of the lib (note that in v6 we have added `exports` entrypoints in the package.json, but the legacy entrypoints from v5 are also still in place; however, the lib is now declared as `module`). This specific `node:crypto` error could be fixed with workarounds, but after that you'll bump into other issues. Really it's not an option to "convert" the Node build to run in browser; we generate platform-specific builds on purpose, for optimization but also security 🙂

Do we have a way to use the browser build using ES import? I have a Next.js TypeScript project. The imports will be
```
import {decrypt, decryptKey, readMessage, readPrivateKey} from 'openpgp';
```
And....yes, this will lead it using node.js build, which would end up causing this error for me.

**Reply by 223740** [reply]
I have pretty much the same situation, we've also added OpenPGP.js via NPM and then try to ES-import it into TypeScript :)

**Reply by twiss** [reply]
As a workaround, you could try importing the browser build file directly, e.g. something like
```js
import { ... } from 'openpgp/dist/openpgp.min.mjs';
```
(untested).

Also.. this is a completely wild guess, but looking at https://github.com/openpgpjs/openpgpjs/blob/121b478312f7bbc86fb6a66f93259a1017c481d7/package.json#L16-L21, it might be better if those values matched exactly; i.e. something like
```json
 "main": "./dist/node/openpgp.min.cjs", 
 "module": "./dist/node/openpgp.min.mjs", 
 "browser": { 
   "./dist/node/openpgp.min.cjs": "./dist/openpgp.min.js", 
   "./dist/node/openpgp.min.mjs": "./dist/openpgp.min.mjs" 
 }, 
```

Would you be able to test whether that fixes things for you?

**Reply by larabr** [reply]
> Do we have a way to use the browser build using ES import? I have a Next.js TypeScript project. The imports will be

Yes. Note that we use the `browser` entrypoint in the `exports` for the browser build, while the `import` one assumes you need the Node one. So you need to configure your bundler/runtime to prefer/use the `browser` condition, if that's not default behaviour. Having a quick look at the [Webpack docs](https://webpack.js.org/configuration/resolve/#resolveconditionnames),  the relevant config option should be e.g.: `resolve: { conditionNames: ['browser', 'import', 'require' ], }`.

If that still doesn't work, importing `openpgp/lightweight` might be the simplest workaround, since that only includes a browser build.

Just for reference, these are our exports from [package.json](https://github.com/openpgpjs/openpgpjs/blob/main/package.json#L22-L33 ) : 
```js
  "exports": {
    ".": {
      "types": "./openpgp.d.ts",
      "import": "./dist/node/openpgp.mjs",
      "require": "./dist/node/openpgp.min.cjs",
      "browser": "./dist/openpgp.min.mjs"
    },
    "./lightweight": {
      "types": "./openpgp.d.ts",
      "browser": "./dist/lightweight/openpgp.min.mjs"
    }
  },
```

**Reply by larabr** [reply]
Update: it seems Webpack enforces the order of the exports as declared by libraries, and `resolve.conditionNames` cannot be set to workaround this behaviour in this specific case 🫤 I've just released a patch (https://github.com/openpgpjs/openpgpjs/commit/f75447afaa681dc6fa7448a4bf82c63b10265b46), can you check if updating to [v6.0.1](https://github.com/openpgpjs/openpgpjs/releases/tag/v6.0.1) fixes the problem?

**Reply by 223740** [reply]
Yes, that fixes the issue, thanks, great work!
